### PR TITLE
Make `Var "_"` not special

### PR DESCRIPTION
**Pull request by ccshan**

### CODE REVIEW

**Comment by ccshan** [comment]
@davidweichiang Oops, I accidentally merged this. Let me know what you think...

File: src/Parse/Parse.hs [hunk 1]
```diff
@@ -159,9 +160,28 @@ parseTerm1 = parsePeeks 2 >>= \ t1t2 -> case t1t2 of
...
+    parseDrop TkIn
+    tm' <- parseTerm1
+    return (UsElimMultiplicative tm xs tm')
 -- let <..., _, x, _, ...> = term in term
-  [TkLet, TkLangle] -> parseEat *> parseEat *> pure (flip (UsElimProd Additive)) <*> parseVarsCommas TkRangle False True <* parseDrop TkEq <*> parseTerm1 <* parseDrop TkIn <*> parseTerm1
+  [TkLet, TkLangle] -> do
+    parseEat
+    parseEat
+    xs <- parseVarsCommas TkRangle False True
+    case [ (i,x) | (i,x) <- enumerate xs, x /= Var "_" ] of
```
Comment (davidweichiang) [inline src/Parse/Parse.hs:177]: Why not push the distinction earlier and make "_" a different Token?

File: src/Struct/Exprs.hs [hunk 2]
```diff
@@ -72,7 +72,8 @@ data UsTm =
   | UsFactor Double UsTm                -- factor wt in tm
   | UsFail Type                         -- fail : tp
   | UsProd AddMult [UsTm]               -- (tm1, ..., tmn)/<tm1, ..., tmn>
-  | UsElimProd AddMult UsTm [Var] UsTm  -- let (x,y,z)/<x,y,z> = tm1 in tm2
+  | UsElimMultiplicative UsTm [Var] UsTm-- let (x,y,z) = tm1 in tm2
+  | UsElimAdditive UsTm Int Int Var UsTm-- let <_,y,_> = tm1 in tm2
```
Comment (davidweichiang) [inline src/Struct/Exprs.hs:76]: could you add a comment explaining the meaning of the two aunts?

Comment (davidweichiang) [reply src/Struct/Exprs.hs:76]: Ints, I mean.

File: src/Struct/Exprs.hs [hunk 3]
```diff
@@ -93,7 +94,8 @@ data Term =
   | TmAmb [Term] Type                             -- amb tm1 tm2 ... tmn : tp
   | TmFactor Double Term Type                     -- factor wt in tm : tp
   | TmProd AddMult [Arg]                          -- (tm1 : tp1, tm2 : tp2, ..., tmn : tpn) / <...>
-  | TmElimProd AddMult Term [Param] Term Type     -- let (x:X,y:Y,z:Z)/<...> = tm1 in tm2 : tp
+  | TmElimMultiplicative Term [Param] Term Type   -- let (x:X,y:Y,z:Z) = tm1 in tm2 : tp
+  | TmElimAdditive Term Int Int Param Term Type   -- let <_..,y:Y,_..> = tm1 in tm2 : tp
```
Comment (davidweichiang) [inline src/Struct/Exprs.hs:98]: Something possibly for a new issue: there's no place for the type of tm1 here, whereas generally the other cases of Term store the types of the subterms.

I think the direction to go in would be to reduce redundancy and not store the types of subterms (they can always be recovered using `typeof`), but was not sure if @HerbertMcSnout had a reason for designing the type like this.

File: src/Parse/Parse.hs (hunk 1 above)
Comment (ccshan) [reply src/Parse/Parse.hs:177]: Oh that's a good point. I think we should continue to allow _ in lambda, case, and ElimMultiplicative though, and I'm not sure the best way to represent that...

Comment (davidweichiang) [reply src/Parse/Parse.hs:177]: Do you want _ to have the meaning "ignore me" or "variable named _"?

**Pull request merged**